Allow overriding the settings based on endpoint

mstriemer opened the pull request:
This allows us to override the CORS settings based on the endpoint. The settings API is a little clunky and I'm not sure if it should be override or different settings entirely but putting this up so we can discuss it.

Supports https://github.com/mozilla/addons-server/pull/2527.

mstriemer committed:
Allow overriding the settings based on endpoint

EnTeQuAk commented:
What I had in mind was something like

``` python
# Default settings
CORS_SETTINGS = {
    '^.*$': {
        'allow_headers': default_headers
        'allow_methods': default_methods
        'allow_credentials': False
        'preflight_max_age': 86400
        'origin_allow_all': False
        # ...
    }
}


# User settings, all default settings are applied to every endpoint mapping and these are only the overwrites
CORS_SETTINGS = {
    '^/api/v3/internal/.*$': {
        'origin_whitelist': ('foob.bar.com', 'bar.foo.com'),
    }
}
```

which is a bit like what django does for `DATABASES` and `CACHES` but to be honest I like your approach maybe a bit more since we're talking about URLs here and yours resembles djangos urlresolver config so I like that :+1:

EnTeQuAk commented on corsheaders/defaults.py:
```diff
@@ -21,28 +21,50 @@
+            'CORS_URLS_REGEX',
+            'CORS_MODEL',
+            'CORS_REPLACE_HTTPS_REFERER',
+        ]
```
you could maybe get that list with

``` python
fields = [field for field in dir(CorsHeadersSettings) if field.startswith('CORS_')]
```

to avoid adding a new config value but not adding it in that fields list.

EnTeQuAk commented on corsheaders/tests.py:
```diff
@@ -183,155 +182,184 @@ def assertAccessControlAllowOriginEquals(self, response, header):
+
+    def test_process_response_in_whitelist(self, get_settings):
+        with override_settings(
+                CORS_MODEL = None, CORS_ORIGIN_ALLOW_ALL = False,
```
whitespaces around `=`, you have that a lot so I'll only comment once :sweat_smile:

EnTeQuAk commented:
I like it, just minor comments, r+wc

diox commented:
WFM. Shall we try to get this upstream ?

mstriemer committed:
Update based on comments

mstriemer committed:
Fix failing test

The old code had settings patched as a MagicMock so
settings.CORS_ALLOW_CREDENTIALS was truthy since it wasn't set and
returned another MagicMock instance.

mstriemer committed:
Remove unused code

mstriemer commented:
I thought about doing something where it would apply all of the overrides that it found but I figured that could turn into a mess if you started adding lots of overrides.

This is formatted like django urls since the regexes need to have an order otherwise you'd get non-deterministic results with a dict:

``` py
CORS_SETTINGS = {
  r'^/api/v3/internal/.*$': { ... },
  r'^/api/v3/.*$': { ... },
}
CORS_SETTINGS.items()  # Which regex is first?
```

mstriemer committed:
Update readme
